### ISSUE TITLE
Several LIGHTING4 commands not supported in RFXCOM binding

### ISSUE THREAD

**Issue by buzink** [openhab/openhab#3858]
I bought some cheap [contact](https://github.com/openhab/openhab/blob/master/bundles/binding/org.openhab.binding.rfxcom/src/main/java/org/openhab/binding/rfxcom/internal/messages/RFXComLighting4Message.java) and motion sensors based on the 1527 chipset. I tried to use them with the RFXCOM binding. Unfortunately the devices do not send an ON command on contact, but a UNKNOWN command. Domoticz treats these commands as ON, so they can be used in that platform without problem (if you set auto update to OFF after a few seconds). Unfortunately, openHAB throws exceptions. I think I tracked the problem down to [this file](https://github.com/openhab/openhab/blob/master/bundles/binding/org.openhab.binding.rfxcom/src/main/java/org/openhab/binding/rfxcom/internal/messages/RFXComLighting4Message.java)

If i interpret the code correctly, all command types of the LIGHTING4 protocol are defined ([line 70](https://github.com/openhab/openhab/blob/master/bundles/binding/org.openhab.binding.rfxcom/src/main/java/org/openhab/binding/rfxcom/internal/messages/RFXComLighting4Message.java#L70)), but only ON/OFF is handled correctly, the rest throw exceptions ([line 190](https://github.com/openhab/openhab/blob/master/bundles/binding/org.openhab.binding.rfxcom/src/main/java/org/openhab/binding/rfxcom/internal/messages/RFXComLighting4Message.java#L190)). Additionally, the RawData type is allowed ([line 93](https://github.com/openhab/openhab/blob/master/bundles/binding/org.openhab.binding.rfxcom/src/main/java/org/openhab/binding/rfxcom/internal/messages/RFXComLighting4Message.java#L93)) but not implemented ([line 232](https://github.com/openhab/openhab/blob/master/bundles/binding/org.openhab.binding.rfxcom/src/main/java/org/openhab/binding/rfxcom/internal/messages/RFXComLighting4Message.java#L93)). 

It would be perfect when:
1. The binding treats the UNDEFINED_0, UNDEFINED_2, UNDEFINED_3 and UNKNOWN commands as ON commands, so the motion and contact sensors can be assigned to Switch and Contact items. (Switch ContactSensor01 { rfxcom="<9287177:Command" })
2. The binding allows for using the RawMode to put the received command into a String (or Selection?) item. (String ContactSensor01 { rfxcom="<9287177:RawData" })

Thanks for looking into this!
##

**Comment by Speederc** [comment]
Hi buzink,

It seems that there are few typos in the RFXComLighting4Message.java that mishandle the RFXCOM messages for Lighting4. In order to see if this correction could help for your problem, can you send me the log file of RFXMngr for ON and OFF commands with your sensor ?

Thanks

**Comment by buzink** [comment]
Hi Speederc,

The sensors give only one command (they are cheap :). This are the RFXMngr readings:

Motion sensor in openhab:

     Raw data = 091300AE8DB699018170
      - Packet type = LIGHTING4
      - Seq number = 174
      - Sub type = PT2262
      - Id = 9287177
      - Command = UNKNOWN
      - Pulse = 385

Motion sensor in RFXMngr:

     Packettype    = Lighting4
     subtype       = PT2262
     Sequence nbr  = 1
     Code          = 8DB699 decimal:9287321
     S1- S24  = 1000 1101 1011 0110 1001 1001
     Pulse         = 385 usec
     Signal level  = 8  -56dBi

Contactsensor in openhab:

     Raw data = 091300B128D109016770
      - Packet type = LIGHTING4
      - Seq number = 177
      - Sub type = PT2262
      - Id = 2674944
      - Command = UNKNOWN
      - Pulse = 359

Contactsensor in RFXMngr:

     Packettype    = Lighting4
     subtype       = PT2262
     Sequence nbr  = 10
     Code          = 28D109 decimal:2674953
     S1- S24  = 0010 1000 1101 0001 0000 1001
     Pulse         = 362 usec
     Signal level  = 7  -64dBi

(I have a few sensors of each, that's why the id's are different)

Thanks!

**Comment by Speederc** [comment]
Hi @buzink ,
I compiled a new version to improve the lighting4 support : http://dr-petelot-marion.chirurgiens-dentistes.fr/temp/org.openhab.binding.rfxcom_1.9.0.201601240042.jar .
According to different examples founded on the web, I did the following mapping : 
   - OFF : 0,2,4,12  
   - ON : 1,3,5
The rawdata string is also added.
Can you test it and give me your feedback ?
Thanks

**Comment by buzink** [comment]
Thanks! I really appreciate it!

The rawdata works now. I can use it as a workaround to make a switch working (if string item x received update then sendcommand ON to switch y). 

But the switch items do not work directly yet. In the log I still see the UNKNOWN command. Would it not be easier to treat everything as ON except the known OFF commands? If for some reason you don't want to go that route, could you explain me how you deduct the command number from the rawdata string? I could test 15 of those sensors and list the command numbers for you if that helps.

**Comment by Speederc** [comment]
I forgot to add the 9 as ON command. This should be fix in http://dr-petelot-marion.chirurgiens-dentistes.fr/temp/org.openhab.binding.rfxcom_1.9.0.201601280033.jar . Can you test it ?
The command ID should be written in the log with this version. I would like to avoid to treat all commands as ON except the known OFF because the logic between ON and OFF IDs is not really clear for now.

**Comment by buzink** [comment]
Really nice that the command ID is written in the log. Strangely enough the command seems to be 9, but is still not recognized as ON command:

```
2016-01-28 21:33:01.815 [DEBUG] [.b.r.internal.RFXComConnection] - Data received:
Raw data = 09130010B2C019017560
 - Packet type = LIGHTING4
 - Seq number = 16
 - Sub type = PT2262
 - Id = 732161
 - Command = UNKNOWN(9)
 - Pulse = 373
2016-01-28 21:33:01.822 [WARN ] [.rfxcom.internal.RFXComBinding] - Data conversion error
org.openhab.binding.rfxcom.internal.RFXComException: Can't convert UNKNOWN to SwitchItem
	at org.openhab.binding.rfxcom.internal.messages.RFXComLighting4Message.convertToState(RFXComLighting4Message.java:227) ~[bundlefile:na]
	at org.openhab.binding.rfxcom.internal.RFXComBinding$MessageLister.packetReceived(RFXComBinding.java:270) ~[bundlefile:na]
	at org.openhab.binding.rfxcom.internal.connector.RFXComSerialConnector$SerialReader.run(RFXComSerialConnector.java:209) [bundlefile:na]
```

**Comment by Speederc** [comment]
Now it should be fine :  http://dr-petelot-marion.chirurgiens-dentistes.fr/temp/org.openhab.binding.rfxcom_1.9.0.201601282303.jar . Sorry for this iterative testing...

**Comment by buzink** [comment]
Yes! That's it. Both the motion and contact sensors work now. Thank you very much. I also understand the commands and codes much better now.

I have a set of outdoor switches from Everflourish that send the commands OFF_12 (switch on) and OFF_4 (switch off). I understand now that both are interpreted as the same OFF by openHAB. Does every manufacturer tread on/off codes differently or could this also be a typo (at least in my case 12 should be ON)?

**Comment by Speederc** [comment]
It was a typo for 12 as other users seem to also received ON messages with 12. I changed it for ON : http://dr-petelot-marion.chirurgiens-dentistes.fr/temp/org.openhab.binding.rfxcom_1.9.0.201601292333.jar . This code will be integrated in a future pull request which should be included in Openhab 1.9.0. Thanks for the testing !

**Comment by buzink** [comment]
Hm, strange. Now I get UNKNOWN(7) instead of OFF_12 for the ON-commands:

```
2016-01-30 01:17:49.429 [DEBUG] [.b.r.internal.RFXComConnection] - Data received:
Raw data = 09130023451557016950
 - Packet type = LIGHTING4
 - Seq number = 35
 - Sub type = PT2262
 - Id = 282965
 - Command = UNKNOWN(7)
 - Pulse = 361

2016-01-30 01:17:50.873 [DEBUG] [.b.r.internal.RFXComConnection] - Data received:
Raw data = 09130026451554016250
 - Packet type = LIGHTING4
 - Seq number = 38
 - Sub type = PT2262
 - Id = 282965
 - Command = OFF_4(4)
 - Pulse = 354

2016-01-30 01:17:52.234 [DEBUG] [.b.r.internal.RFXComConnection] - Data received:
Raw data = 09130027454557016560
 - Packet type = LIGHTING4
 - Seq number = 39
 - Sub type = PT2262
 - Id = 283733
 - Command = UNKNOWN(7)
 - Pulse = 357

2016-01-30 01:17:53.186 [DEBUG] [.b.r.internal.RFXComConnection] - Data received:
Raw data = 09130029454554016260
 - Packet type = LIGHTING4
 - Seq number = 41
 - Sub type = PT2262
 - Id = 283733
 - Command = OFF_4(4)
 - Pulse = 354

2016-01-30 01:17:53.937 [DEBUG] [.b.r.internal.RFXComConnection] - Data received:
Raw data = 0913002B455157016560
 - Packet type = LIGHTING4
 - Seq number = 43
 - Sub type = PT2262
 - Id = 283925
 - Command = UNKNOWN(7)
 - Pulse = 357

2016-01-30 01:17:54.850 [DEBUG] [.b.r.internal.RFXComConnection] - Data received:
Raw data = 0913002C455154016260
 - Packet type = LIGHTING4
 - Seq number = 44
 - Sub type = PT2262
 - Id = 283925
 - Command = OFF_4(4)
 - Pulse = 354

```

**Comment by Speederc** [comment]
Last version with 7 also as ON : http://dr-petelot-marion.chirurgiens-dentistes.fr/temp/org.openhab.binding.rfxcom_1.9.0.201601301041.jar
To resume, the following mapping between commandID and command is used :
-     OFF : 0,2,4
-     ON : 1,3,5,7,12

**Comment by buzink** [comment]
It works! Thank you very much for all your effort. I am really happy that all my sensors and remotes work now.If you ever need some testing done, feel free to contact me.

**Comment by buzink** [comment]
For reference to others: the cheap contact sensors, motion sensors and outdoor switches talked about in this topic are:

- [Contact sensors](http://www.ebay.nl/itm/181185686979?_trksid=p2060353.m2749.l2649&var=480260705755&ssPageName=STRK%3AMEBIDX%3AIT) and [motion sensors](http://www.ebay.nl/itm/171089657359?_trksid=p2060353.m2749.l2649&var=470275442985&ssPageName=STRK%3AMEBIDX%3AIT) with a 1507 chip set directly bought in China through ebay/banggood/aliexpress.com for 8 and 4 dollars a piece.
- Cheap EMW200T outdoor switches from EverFlourish/Maxeda DIY for sale in European shops.

Both use the LIGHTING4 protocol and can be used with openHAB through the RFXcom transceiver. The motion and contact sensors only know a ON state, useful for security, worthless for knowing the open/closed state of a window.

**Comment by teichsta** [comment]
are you going to send this changes as PR @Speederc?

**Comment by Speederc** [comment]
I will create different PRs for all these RFXCOM updates (fixes, new messages type, new sensors).

**Comment by sheppy99** [comment]
I've just got hold of a Aoke 4 relay board which needs the full range of values sending to it from 0 to 15, to select all combinations of the 4 relays, is there a way of configuring the modified lighting4 protocol to do this? Currently the switch item seems to send either 1 or 4 in response to ON and OFF (can't remember which order)

**Comment by 9037568** [comment]
Closed via merged PR #3953
